git/types.py

```diff
@@ -33,85 +33,165 @@
         runtime_checkable,
     )
 
-# if sys.version_info >= (3, 10):
-#     from typing import TypeGuard  # noqa: F401
-# else:
-#     from typing_extensions import TypeGuard  # noqa: F401
-
-PathLike = Union[str, "os.PathLike[str]"]
-
 if TYPE_CHECKING:
     from git.repo import Repo
     from git.objects import Commit, Tree, TagObject, Blob
 
-    # from git.refs import SymbolicReference
+PathLike = Union[str, "os.PathLike[str]"]
+"""A :class:`str` (Unicode) based file or directory path."""
 
 TBD = Any
+"""Alias of :class:`~typing.Any`, when a type hint is meant to become more specific."""
+
 _T = TypeVar("_T")
+"""Type variable used internally in GitPython."""
 
 Tree_ish = Union["Commit", "Tree"]
+"""Union of :class:`~git.objects.base.Object`-based types that are inherently tree-ish.
+
+See gitglossary(7) on "tree-ish": https://git-scm.com/docs/gitglossary#def_tree-ish
+
+:note:
+    This union comprises **only** the :class:`~git.objects.commit.Commit` and
+    :class:`~git.objects.tree.Tree` classes, **all** of whose instances are tree-ish.
+    This is done because of the way GitPython uses it as a static type annotation.
+
+    :class:`~git.objects.tag.TagObject`, some but not all of whose instances are
+    tree-ish (those representing git tag objects that ultimately resolve to a tree or
+    commit), is not covered as part of this union type.
+"""
+
 Commit_ish = Union["Commit", "TagObject", "Blob", "Tree"]
+"""Union of the :class:`~git.objects.base.Object`-based types that represent git object
+types. This union is often usable where a commit-ish is expected, but is not actually
+limited to types representing commit-ish git objects.
+
+See gitglossary(7) on:
+
+* "commit-ish": https://git-scm.com/docs/gitglossary#def_commit-ish
+* "object type": https://git-scm.com/docs/gitglossary#def_object_type
+
+:note:
+    This union comprises **more** classes than those whose instances really represent
+    commit-ish git objects:
+
+    * A :class:`~git.objects.commit.Commit` is of course always commit-ish, and a
+      :class:`~git.objects.tag.TagObject` is commit-ish if, when peeled (recursively
+      followed), a :class:`~git.objects.commit.Commit` is obtained.
+    * However, :class:`~git.objects.blob.Blob` and :class:`~git.objects.tree.Tree` are
+      also included, and they represent git objects that are never really commit-ish.
+
+    This is an inversion of the situation with :class:`Tree_ish`, which is narrower than
+    all tree-ish objects. It is done for practical reasons including backward
+    compatibility.
```

Comment:
This is the second of two places in the `Commit_ish` docstring where I have said something that sounds plausible but may be inaccurate. The last sentence here makes two vague claims, which even if vague should only be kept if true (or maybe removed even then, if not useful), and which preferably should be made less vague.